File: python/enolib/error_types.py
class Error(Exception):
    def __init__(self, text, snippet, selection):
        super().__init__(f"{text}\n\n{snippet}")

        self.cursor = selection['from']
        self.message = f"{text}\n\n{snippet}"
        self.selection = selection
        self.snippet = snippet
        self.text = text


class ParseError(Error):
    pass


class ValidationError(Error):
    pass

File: javascript/src/error_types.ts
export class EnoError extends Error {
    constructor(text, snippet, selection) {
        super(`${text}\n\n${snippet}`);
        
        this.selection = selection;
        this.snippet = snippet;
        this.text = text;
        
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, EnoError);
        }
    }
    
    get cursor() {
        return this.selection.from;
    }
}

export class ParseError extends EnoError {
    constructor(...args) {
        super(...args);
        
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, ParseError);
        }
    }
}

export class ValidationError extends EnoError {
    constructor(...args) {
        super(...args);
        
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, ValidationError);
        }
    }
}
